Generalizing cross validation for other data types and Trainers.

crb3874 opened the pull request:
Refs #21399

This PR builds on the recent PR for implementing cross-validation capabilities in SurrogateTrainers (https://github.com/idaholab/moose/pull/21405). The main purpose is generalizing for other `Trainers`, and other types of input/output data.

- Refactored `SurrogateTrainer` and several derived classes to handle Reporter gathering for common (`Real`, `std::vector<Real>`) response and predictor types in the base class. This removed some duplicated code in other Trainers in the module.
- Added some API functions and protected member variables in `SurrogateTrainer` for accessing common types of training data in derived Trainers.
- Refactored several Trainers to make them compatible with cross validation.
- Added `std::vector<Real>` response type support to `NearestPointTrainer` (this isn't particularly important, I implemented this for an earlier, now un-used, approach to handling the other points in this PR. I figured I'd leave it in, but it can easily be removed).
- Added documentation for the changes to the `SurrogateTrainer` syntax.
- Added tests for CV with all trainer types, and for CV with `std::vector<Real>` response types.

crb3874 committed:
Generalizing cross validation for other data types and Trainers. Refs #21399

zachmprince commented on modules/stochastic_tools/test/src/reporters/CrossValidationScores.C:
```diff
@@ -18,24 +18,33 @@ CrossValidationScores::validParams()
 CrossValidationScores::execute()
 {
+  unsigned int n_r = _response_indices.size();
   for (unsigned int m = 0; m < _models.size(); ++m)
-    (*_cv_scores[m]) = _models[m]->getModelData<std::vector<Real>>("cv_scores");
+    for (unsigned int r = 0; r < n_r; ++r)
+      for (auto trial : _models[m]->getModelData<std::vector<std::vector<Real>>>("cv_scores"))
+        (*_cv_scores[n_r * m + r]).push_back(trial[_response_indices[r]]);
```
```cpp
  for (unsigned int m = 0; m < _models.size(); ++m)
    (*_cv_scores[m]) = _models[m]->getModelData<std::vector<std::vector<Real>>>("cv_scores");
```

zachmprince commented on modules/stochastic_tools/src/surrogates/NearestPointTrainer.C:
```diff
@@ -17,70 +17,55 @@ NearestPointTrainer::validParams()
+  // Get responses
+  if (_rval)
+    _sample_results[0].push_back(*_rval);
+
```
Remove empty line

zachmprince commented on modules/stochastic_tools/src/surrogates/NearestPointTrainer.C:
```diff
@@ -17,70 +17,55 @@ NearestPointTrainer::validParams()
+    it.clear();
+
+  for (auto & it : _sample_results)
+    it.clear();
```
Should we be doing a reserve here too?

zachmprince commented on modules/stochastic_tools/src/surrogates/NearestPointSurrogate.C:
```diff
@@ -22,19 +22,37 @@ NearestPointSurrogate::validParams()
+}
+
+Real
+NearestPointSurrogate::findNearestPoint(const std::vector<Real> & x, const unsigned int r) const
```
Let's make this slightly more efficient. Instead of returning the value, we should return the index in `_sample_points` then the `evaluate` functions can just use it to into `_sample_results`.

zachmprince commented on modules/stochastic_tools/examples/surrogates/cross_validation/gold/normal_train_and_cv_out_cv_scores_0002.csv:
```diff
@@ -1,4 +1,4 @@
-pr_surr
+pr_surr_0
```
Do we need these polynomial regression-only tests anymore?

zachmprince reviewed (changes requested):
Very simple stuff here. Great work!

crb3874 committed:
Addressing review comments for generalizing CV options - refs #21565

crb3874 commented:
Sorry this took so long to get addressed - I'm glad we decided to go through the analysis with all trainer types to add to the documentation, but it took a little while - the main limitation was that the configuration I'd used before (1e5 trials with 6560 sampler rows) was significantly more expensive to run with GP and ANN, and was probably overkill to begin with - I reduced the problem to 100 repetitions/1000 sampler rows to make it tractable. I only included results with uniform parameters because the results with normal parameters wasn't very interesting.

I also had issues with accuracy/convergence with the neural network. I got some help from Peter on setting better input parameters, and went ahead and implemented data standardization for ANN (not included in this PR) - this improved the results I was getting significantly. 

Otherwise, I believe I addressed all the requested changes. I removed the regression tests for PR-only, but left the input files in the examples folder since they're referred to in the documentation. I also re-golded as necessary to handle the change to JSON output for CrossValidationScores. Otherwise, I agree with all the changes/they're in this recent commit.

Thanks!

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1114966/) on 7796c10 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21565/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1116024/) on 59b61e8 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Stochastic tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>b2ec4f</th>
      <th colspan=4>#21565 59b61e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.07%</td>
      <td>90.05%</td>
      <td>-0.02%</td>
      <td>89.15%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4970</td>
      <td>4995</td>
      <td>+25</td>
      <td>115</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>548</td>
      <td>552</td>
      <td>+4</td>
      <td>14</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21565/coverage_diff/stochastic_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21565/coverage/stochastic_tools)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21565/coverage/xfem)

</details>

## Warnings

- `stochastic_tools` new line coverage rate 89.15% is less than the suggested 90.0%



This comment will be updated on new commits.
